Fix up MDN API landing page template for browser-compat

hamishwillee opened the pull request:
The new way of doing browser compatibility sections is to list the BCD path in the page frontmatter `browser-compat` key and just use `{{Compat}}` as a placeholder. Also we no longer have hidden sections.

This fixes just one template: https://developer.mozilla.org/en-US/docs/MDN/Structures/Page_types/API_landing_page_template. If this is merged I'll do the other MDN instructional docs/templates - e.g. https://developer.mozilla.org/en-US/docs/MDN/Structures/Page_types/SVG_element_page_template

This follows follows on naturally from https://github.com/mdn/content/pull/4398 and https://github.com/mdn/content/issues/2228

@wbamberg @Elchi3 Sanity check would be good :-)

hamishwillee committed:
API landing page template: Fix compat

typo

wbamberg commented:
Thanks for doing this @hamishwillee ! Definitely needed, and this general update looks fine. Having said that, I'm not actually sure yet whether we will be doing this for API pages in particular. See https://github.com/mdn/content/discussions/4920.

What you've written here definitely applies for https://developer.mozilla.org/en-US/docs/MDN/Structures/Page_types#css_feature_reference_page and very likely:

https://developer.mozilla.org/en-US/docs/MDN/Structures/Page_types#api_reference_subpage
https://developer.mozilla.org/en-US/docs/MDN/Structures/Page_types#html_element_reference_page
https://developer.mozilla.org/en-US/docs/MDN/Structures/Page_types#svg_element_reference_page
https://developer.mozilla.org/en-US/docs/MDN/Structures/Page_types#http_header_reference_page

...and of course JS, which does not seem to be documented here... 

This update could also talk about using the {{Specifications}} macro, instead of a handwritten table.

hamishwillee commented:
Hi @wbamberg,  Thank you for the feedback.  IMO there isn't any reason not to roll this out to API docs and everywhere else. Even if we are sporadic about roll out, the howto docs should reflect a consistent approach. 

I will add update for `{{Specifications}}` too. Originally I decided not to because these are not as well rolled out. But then I realised that's not the point as this is for new docs, and the update to BCD needs to include this information going forward.

hamishwillee committed:
Update with Specifications macro

hamishwillee committed:
Fix up master to main in the compatibility table info

hamishwillee committed:
Fix up AbortController - better example

hamishwillee committed:
Update instructions on how to work with compatibility tables

hamishwillee committed:
minor fixes to api landing page

hamishwillee commented:
@wbamberg et al, OK I did a further update to add info on how to update the specifications table.

That lead me to wondering how you update the `spec_url`, which lead to me updating https://developer.mozilla.org/en-US/docs/MDN/Structures/Compatibility_tables with:
- basic information on how to find the spec URL 
- Fixes to update from pointing to git main branch rather than master
- Updated AbortController as an example (it is a better example than the other two as it has both the spec and browser compat.

Anyway, needs a re-review before I go and roll everything out everywhere :-)

hamishwillee commented on files/en-us/mdn/structures/compatibility_tables/index.html:
```diff
@@ -446,17 +450,38 @@ <h2 id="Pushing_a_change_back_to_the_main_repo">Pushing a change back to the mai
+<p>The macro calls generate the following tables (and corresponding set of notes):</p>
+
+
+<h4 id="Compatibility_table_example">Compatibility table example</h4>
```
Note, headings needed to force the tables below the things that talk about them.

hamishwillee commented on files/en-us/mdn/structures/compatibility_tables/index.html:
```diff
@@ -446,17 +450,38 @@ <h2 id="Pushing_a_change_back_to_the_main_repo">Pushing a change back to the mai
-<div class="notecard note">
-<p><strong>Note:</strong> The filenames often match the labels given to the interfaces inside the JSON structures, but it is not always the case. When the macro calls generate the tables, they walk through all the files until they find the relevant JSON to use, so the filenames are not critical. Saying that, you should always name them as intuitively as possible.</p>
-</div>
+<p>You can also specify the desired API as the first argument to the macro as shown: <code>\{{Compat("api.AbortController")}}</code>. This can be useful if multiple compatibility tables are required on the same page.</p>
```
FWIW I don't know if this is something we still want to advertise.

Elchi3 replied:
Yep, this is fine. Guide pages typically want a compat table (or multiple).

Elchi3 commented on files/en-us/mdn/structures/compatibility_tables/index.html:
```diff
@@ -446,17 +450,38 @@ <h2 id="Pushing_a_change_back_to_the_main_repo">Pushing a change back to the mai
-<p>Once your new data has been included in the main repo, you can start dynamically generating browser compat tables based on that data on MDN pages using the {{TemplateLink("Compat")}} macro. This takes a single parameter, the dot notation required to walk down the JSON data and find the object representing the feature you want to generate the compat table for.</p>
+<p>Once your new compatibility data has been included in the main <a href="https://github.com/mdn/browser-compat-data">browser-compat-data</a> repo, you can start dynamically generating browser compatibility and specification tables based on that data on MDN pages.</p>
+
+<p>First you need to find the "address" for the relevant compatibility data you wish to include. This can be determined by inspecting its source file. For example:</p>
```
I think we've been calling this query string or just query in other places.

```suggestion
<p>First you need to find the query string for the relevant compatibility data you wish to include. This can be determined by inspecting its source file. For example:</p>
```

Elchi3 commented on files/en-us/mdn/structures/compatibility_tables/index.html:
```diff
@@ -446,17 +450,38 @@ <h2 id="Pushing_a_change_back_to_the_main_repo">Pushing a change back to the mai
+
+<p>First you need to find the "address" for the relevant compatibility data you wish to include. This can be determined by inspecting its source file. For example:</p>
+<ul>
+  <li>{{domxref("AbortController")}} compatibility data is defined in <a href="https://github.com/mdn/browser-compat-data/blob/main/api/AbortController.json">AbortController.json</a> and has address <code>api.AbortController</code>.</li>
```
```suggestion
  <li>{{domxref("AbortController")}} compatibility data is defined in <a href="https://github.com/mdn/browser-compat-data/blob/main/api/AbortController.json">AbortController.json</a> and can be queried using <code>api.AbortController</code>.</li>
```

Elchi3 commented on files/en-us/mdn/structures/compatibility_tables/index.html:
```diff
@@ -446,17 +450,38 @@ <h2 id="Pushing_a_change_back_to_the_main_repo">Pushing a change back to the mai
+<p>First you need to find the "address" for the relevant compatibility data you wish to include. This can be determined by inspecting its source file. For example:</p>
+<ul>
+  <li>{{domxref("AbortController")}} compatibility data is defined in <a href="https://github.com/mdn/browser-compat-data/blob/main/api/AbortController.json">AbortController.json</a> and has address <code>api.AbortController</code>.</li>
+  <li>{{HTTPHeader("Content-Type")}} HTTP header compatibility data is defined in <a href="https://github.com/mdn/browser-compat-data/blob/main/http/headers/content-type.json">content-type.json</a> and has address <code>http.headers.Content-Type</code>.</li>
```
```suggestion
  <li>{{HTTPHeader("Content-Type")}} HTTP header compatibility data is defined in <a href="https://github.com/mdn/browser-compat-data/blob/main/http/headers/content-type.json">content-type.json</a> and the query is <code>http.headers.Content-Type</code>.</li>
```

Elchi3 commented on files/en-us/mdn/structures/compatibility_tables/index.html:
```diff
@@ -446,17 +450,38 @@ <h2 id="Pushing_a_change_back_to_the_main_repo">Pushing a change back to the mai
+<ul>
+  <li>{{domxref("AbortController")}} compatibility data is defined in <a href="https://github.com/mdn/browser-compat-data/blob/main/api/AbortController.json">AbortController.json</a> and has address <code>api.AbortController</code>.</li>
+  <li>{{HTTPHeader("Content-Type")}} HTTP header compatibility data is defined in <a href="https://github.com/mdn/browser-compat-data/blob/main/http/headers/content-type.json">content-type.json</a> and has address <code>http.headers.Content-Type</code>.</li>
+  <li>{{domxref("VRDisplay.capabilities")}} property compatibility data is defined in <a href="https://github.com/mdn/browser-compat-data/blob/main/api/VRDisplay.json">VRDisplay.json</a> and has address key is <code>api.VRDisplay.capabilities</code>.</li>
```
```suggestion
  <li>{{domxref("VRDisplay.capabilities")}} property compatibility data is defined in <a href="https://github.com/mdn/browser-compat-data/blob/main/api/VRDisplay.json">VRDisplay.json</a> and its query is <code>api.VRDisplay.capabilities</code>.</li>
```

Elchi3 commented on files/en-us/mdn/structures/compatibility_tables/index.html:
```diff
@@ -446,17 +450,38 @@ <h2 id="Pushing_a_change_back_to_the_main_repo">Pushing a change back to the mai
+</ul>
 
-<p>As an example, on the {{HTTPHeader("Content-Type")}} HTTP header page, the macro call looks like this: <code>\{{Compat("http.headers.Content-Type")}}</code>. If you look at the <a href="https://github.com/mdn/browser-compat-data/blob/main/http/headers/content-type.json">content-type.json</a> file in the repo, you'll see how this is reflected in the JSON data.</p>
+<p>The compatibility data address should be specified in the page front-matter using the <code>browser-compat</code> key.
```
```suggestion
<p>The compatibility data query should be specified in the page front-matter using the <code>browser-compat</code> key.
```

Elchi3 reviewed (dismissed):
Thanks for updating this!

wbamberg commented:
> Hi @wbamberg, Thank you for the feedback. IMO there isn't any reason not to roll this out to API docs and everywhere else. 

For *API landing pages* specifically, it's not at all obvious to me that the approach of adding a single BCD query into front matter makes sense, because the tech covered by a single API landing page doesn't correspond to a single BCD query. For example, which single query would you use to describe the compat status of the WebRTC API: https://developer.mozilla.org/en-US/docs/Web/API/WebRTC_API ?

hamishwillee committed:
Update files/en-us/mdn/structures/compatibility_tables/index.html

Co-authored-by: Florian Scholz <fs@florianscholz.com>

Update files/en-us/mdn/structures/compatibility_tables/index.html

Co-authored-by: Florian Scholz <fs@florianscholz.com>

Update files/en-us/mdn/structures/compatibility_tables/index.html

Co-authored-by: Florian Scholz <fs@florianscholz.com>

Update files/en-us/mdn/structures/compatibility_tables/index.html

Co-authored-by: Florian Scholz <fs@florianscholz.com>

Update files/en-us/mdn/structures/compatibility_tables/index.html

Co-authored-by: Florian Scholz <fs@florianscholz.com>

hamishwillee committed:
Method_subpage_mimimal

hamishwillee committed:
Updated guide pattern for api method subpage

hamishwillee committed:
Add API reference page template

hamishwillee committed:
Update/partially revert landing page

hamishwillee commented:
@wbamberg @Elchi3 

Re https://github.com/mdn/content/pull/5225#issuecomment-848426600 "For API landing pages specifically, it's not at all obvious ..."

OK, I get you. Part of the problem is that it is not all that clear in the original doc. At the moment it talks about just one compat table. What I **think** this should say is that you may find it useful to display BCD tables for the main interfaces in an API - so work out the keys for each of them and put them in compatibility section using separate  compat macro (or something similar). Does that make sense? For now I have just reverted that section.

If so, what about specifications? That macro does not take a BCD entry. So do we say "create your own table" or do I remove this section, or what?

FURTHER ...
I have updated the API reference and API reference method pages, because those actually can use the compat and specification macros. As part of that I reworked the top note to better highlight frontmatter. IMO this is much cleaner.

Elchi3 commented on files/en-us/mdn/structures/compatibility_tables/index.html:
```diff
@@ -446,15 +450,38 @@ <h2 id="Pushing_a_change_back_to_the_main_repo">Pushing a change back to the mai
+
+<p>The compatibility and specification tables corresponding to the key are then automatically rendered in place of the <code>\{{Compat}}</code> and <code>\{{Specifications}}</code> macros in the source.</p>
+
+<p>You can also specify the desired API as the first argument to the macro as shown: <code>\{{Compat("api.AbortController")}}</code>. This can be useful if multiple compatibility tables are required on the same page.</p>
```
fwiw, this is also true for specification tables, see https://github.com/mdn/yari/blob/main/kumascript/macros/Specifications.ejs

hamishwillee replied:
Thanks @Elchi3. I did test that and it didn't work. Must have made a mistake. Will try again.

Elchi3 commented:
> If so, what about specifications? That macro does not take a BCD entry. So do we say "create your own table" or do I remove this section, or what?

The specification macro can actually take a BCD query if needed. I think having a manual spec table is fine as the exception, too, though. I mean it is just links to specs really. If certain pages need something more fancy, then I would decide on a case by case basis. The majority of reference-type pages should be fine with just {{Specifications}} imo.

hamishwillee commented:
>> If so, what about specifications? That macro does not take a BCD entry. So do we say "create your own table" or do I remove this section, or what?

> The specification macro can actually take a BCD query if needed. I think having a manual spec table is fine as the exception, too, though. I mean it is just links to specs really. If certain pages need something more fancy, then I would decide on a case by case basis. The majority of reference-type pages should be fine with just {{Specifications}} imo.

Great news. I did test this, which is why I thought it did not work. Will retest and update appropriately.

hamishwillee committed:
Api landing - clarify that multiple specs/compat might be needed

hamishwillee commented:
@Elchi3 I've rebased to master and modified the API Landing page to improve the browser compat and specifications sections as discussed. It now says
- Both these sections are optional
- They can have multiple tables or none. Do what is sensible for the API.

Can you please review and (ideally) merge. After I'll roll out more broadly.

Elchi3 reviewed (approved):
Awesome, thanks Hamish!
